### DISCUSSION TITLE
e2e: organize reusable cy commands with nested functions

### DISCUSSION THREAD

**Discussion by GeenCo86** [cypress-io/cypress#27278] [Questions and Help]
**cypress-version:  12.17.0
ui library: react**

Hello,

I have multiple pages with table components. Now I would like to test every page and use the nested components of the table.
So click on an add button, click on a delete button and so on. To not repeat this code every time, I would like to create a chainable command with nested custom commands.

Something like: 

```
getTable(testId: string): Chainable<{ 
    addButton: () => Chainable<JQuery<HTMLElement>>;
}>
```

```
Cypress.Commands.add('getTable', (selector, ...args) => cy.get(selector, ...args).then((x) => ({
    addButton: () => cy.get('.addButton')
})));
```

and to call it: 

```
cy.getTable('table-test-id').addButtton().click();
```

But I don' find a way to do this. Is there the possibility for something like that?

**Comment by portyu9** [comment]
Hi @GeenCo86 

Yes, you can get essentially that chaining syntax, but not by returning an object containing nested functions.

Cypress chain methods come from the Cypress command registry. For:

```js
cy.getTable('table-test-id').addButton().click()
```

make `addButton()` a **child command** using `prevSubject`:

```js
Cypress.Commands.add('getTable', (testId) => {
  return cy.get(`[data-cy="${testId}"]`)
})

Cypress.Commands.add(
  'addButton',
  { prevSubject: 'element' },
  ($table) => {
    return cy.wrap($table).find('.addButton')
  },
)
```

Then this works naturally:

```js
cy.getTable('table-test-id')
  .addButton()
  .click()
```

`prevSubject: 'element'` means Cypress passes the element yielded by `getTable()` into `addButton()` and validates that the subject is an element.

That said, if these methods are only aliases for selectors, Cypress currently recommends keeping abstractions simple. A normal helper can often be clearer:

```js
const getTable = (testId) => cy.get(`[data-cy="${testId}"]`)

getTable('table-test-id')
  .find('.addButton')
  .click()
```

If you build a reusable DOM query where Cypress's built-in retry behavior is important, also look at `Cypress.Commands.addQuery()` rather than wrapping every selector in a custom command.

The full discussion was reviewed; there were no screenshots/attachments or replies.

Please mark this answer as accepted if it helped, thank you.
